cartocrow/isoline_simplification/isoline_simplifier.cpp

```diff
@@ -0,0 +1,1511 @@
+		if (site.is_point()) {
+			m_p_vertex[site.point()] = vit;
+		} else {
+			m_e_vertex[site.segment()] = vit;
+		}
+	}
+}
+
+bool IsolineSimplifier::simplify(int target, bool debug) {
+	while (m_current_complexity > target) {
+		if (debug && m_current_complexity % 1000 == 0) {
+			std::cout << "\r#Vertices: " << m_current_complexity << std::flush;
+		}
+		if (!step()) return false;
+		update_matching();
+		update_ladders();
+	}
+	return true;
+}
+
+bool IsolineSimplifier::dyken_simplify(int target, double sep_dist) {
+	m_started = true;
+	int start_complexity = m_current_complexity;
+
+	CT ct;
+
+	std::vector<CT::Constraint_id> ids;
+
+	for (const auto& isoline : m_simplified_isolines) {
+		if (isoline.m_closed) {
+			CT::Constraint_id id = ct.insert_constraint(isoline.polygon());
+			ids.push_back(id);
+		} else {
+			CT::Constraint_id id = ct.insert_constraint(isoline.m_points);
+			ids.push_back(id);
+		}
+	}
+
+	std::vector<Isoline<K>> result(m_simplified_isolines.size());
+
+	m_current_complexity -= PS::simplify(ct, Cost(sep_dist), Stop(target));
+	for(auto cit = ct.constraints_begin(); cit != ct.constraints_end(); ++cit) {
+		CT::Constraint_id id = *cit;
+		auto i = std::distance(ids.begin(), std::find(ids.begin(), ids.end(), id));
+
+		std::vector<Point<K>> simplified_points;
+		for (auto vit = ct.points_in_constraint_begin(*cit); vit != ct.points_in_constraint_end(*cit); ++vit) {
+			simplified_points.push_back(*vit);
+		}
+		if (m_simplified_isolines[i].m_closed) {
+			simplified_points.pop_back();
+		}
+		result[i] = Isoline<K>(simplified_points, m_simplified_isolines[i].m_closed);
+	}
+
+	m_simplified_isolines = result;
+
+	return start_complexity != m_current_complexity;
+}
+
+void IsolineSimplifier::collapse_ladder(SlopeLadder& ladder) {
+	auto insert_adj = [this](SDG2::Vertex_handle vertex) {
+		// the vertices that are inserted can probably be reduced / limited by using the fact that
+		// adjacent vertices of the same isoline do not affect the slope ladders.
+		auto ic_start = m_delaunay.incident_vertices(vertex);
+		auto ic = ic_start;
+		if (ic != nullptr) {
+			do {
+				if (ic->storage_site().is_defined()) {
+					m_changed_vertices.insert(ic);
+				} else {
+//					std::cout << "Encountered undefined storage site" << std::endl;
+				}
+			} while (++ic != ic_start);
+		}
+	};
+
+	auto delaunay_remove_p = [&insert_adj, this](const Point<K>& p) {
+	 	auto vertex = m_p_vertex.at(p);
+	 	insert_adj(vertex);
+	  	m_changed_vertices.erase(vertex);
+	  	m_deleted_points.push_back(p);
+
+	    if (!m_delaunay.remove(vertex)) {
+			std::cerr << "\nPoint removal failed\nLikely point " << p << " is incident to a segment that has not yet been deleted." << std::endl;
+			std::cerr << "Rebuilding delaunay triangulation..." << std::endl;
+
+			initialize_sdg();
+			m_deleted_points.clear();
+			m_changed_vertices.clear();
+			return false;
+
+//		    throw std::runtime_error("Delaunay point vertex removal failed!");
```

Comment:
May be good to clean this up before merging.

Reply:
Done.